src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs

```diff
@@ -1062,41 +1062,77 @@ internal static void EmitArray(this ILGenerator il, Type arrayType)
 
         #region Support for emitting constants
 
-        internal static void EmitDecimal(this ILGenerator il, decimal value)
+        private static void EmitDecimal(this ILGenerator il, decimal value)
         {
-            if (decimal.Truncate(value) == value)
+            int[] bits = decimal.GetBits(value);
+            int scale = (bits[3] & int.MaxValue) >> 16;
+            if (scale == 0)
             {
-                if (int.MinValue <= value && value <= int.MaxValue)
+                if (int.MinValue <= value)
                 {
-                    int intValue = decimal.ToInt32(value);
-                    il.EmitPrimitive(intValue);
-                    il.EmitNew(Decimal_Ctor_Int32);
+                    if (value <= int.MaxValue)
+                    {
+                        int intValue = decimal.ToInt32(value);
+                        switch (intValue)
+                        {
+                            case -1:
+                                il.Emit(OpCodes.Ldsfld, Decimal_MinusOne);
+                                return;
+                            case 0:
+                                il.EmitDefault(typeof(decimal));
```

Comment:
@VSadov @JonHanna This new usage of EmitDefault caused a build break since it merged right after https://github.com/dotnet/corefx/pull/15946 which added a new parameter to this method. Can you please take a look and fix this break?